A single failing tag query no longer makes all tag attribution fail

Dzol opened the pull request:
Discussion can be found on the [ticket](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/issues/15#issue-103569940).

Commit:
Test a less trivial pattern for tagged queries

As things stand, rather than a boolean value, this (slightly more
complex) query will return an error tuple. Consequently, NO tags are
attributed to the user by the LDAP plugin, even those tags that have
NOTHING to do with the tag in question.

Commit:
Discard error tuples in the result of a tag query

A single `{error, ...}` meant that a user wouldn't have any tags
attributed to them. Their authorization would fail entirely if any one
of the tag queries failed with an error tuple.

michaelklishin commented:
Not assigning anyone as this will be reviewed by @dcorbacho, @hairyhum, and me.

dcorbacho commented:
I'd like to see a test explicitly named after this bug, instead of affecting the whole test suite. That is, the rabbit config dynamically changed for that test case.

As it is now, the config could be changed in the future to test something else and go unnoticed. Thus, losing the test coverage.

michaelklishin commented:
We can also introduce a new user in the existing config, at least for now.

Dzol commented:
Okay, 

So, @dcorbacho and @michaelklishin and @hairyhum, our two options for moving the necessary configuration out of the file and into the test-set itself are:
1. write a test case from scratch just for tag queries (to inclued the corresponding dynamic configuration changes)
2. change set-up and tear-down for the existing test cases

In the former it'd be more clear that we're testing tag query specifically but it'd lean away from the system-like tests that this suite currently has. In the later it wouldn't be as clear that the set-up and tear-down we'd introduce is for testing tag queries specifically.

@michaelklishin: Is the new user for testing the various success/failure cases for tag queries?

michaelklishin commented:
Option 1 sounds good to me. Using a separate user makes sense for me.

Dzol commented:
Great: proceeding with that then.

Commit:
Test tag attribution by configuring from within the test-suite

Put the necessary configuration in the test itself, testing a more
complex assortment of tag queries, rather than in the configuration
file. Introduce another user into the LDAP server for this
test. Revert the `management` tag query, in the configuration file, to
a trivial constant query.

Dzol commented:
@michaelklishin, @dcorbacho, and @hairyhum: I've made the changes discussed above. One thing we could do, instead of just ignoring failed tag queries, is to log them as a helpful debugging aid.

hairyhum commented:
@Dzol query errors should already be logged to ldap log during execution.

Dzol commented:
@hairyhum: You're right. The logs have plenty of helpful information
about the error at every step of the query which will be much better
for debugging a query than just listing the errored tag queries.

Our concern was that a broken query which results in an error is
interpreted as a FALSE, so any errors will only get noticed if someone
looks for `{error, ...}` in the log. Someone might find that they're
mysteriously not authorized for something when in fact the query is
broken.

michaelklishin commented:
@Dzol that's when it's time to turn on more verbose logging. Logging authorization steps can put a lot of stress on the logger because it would happen every N messages, so I'd leave things as is for now.

Dzol commented:
@michaelklishin: Okay, sure. So failed tag queries are interpreted as a false value (meaning a client won't be attributed the tag), there's a corresponding test case, and logging is more than adequate for debugging a broken query.

Dzol commented:
Currently we only test against LDAP authentication/authorization [here](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/rabbitmq-auth-backend-ldap-15/test/src/rabbit_auth_backend_ldap_test.erl#L108). We'd like to bring this in line with the other system-like tests, before this is merged, with a test generator across the three backend configurations.

michaelklishin commented:
@Dzol what three backend configurations and what other tests?

Dzol commented:
@michaelklishin: I meant the backend configurations that correspond to [`ldap_only_test_`](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/rabbitmq-auth-backend-ldap-15/test/src/rabbit_auth_backend_ldap_test.erl#L36), [`ldap_and_internal_test_`](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/rabbitmq-auth-backend-ldap-15/test/src/rabbit_auth_backend_ldap_test.erl#L49), and [`internal_followed_ldap_and_internal_test_`](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/rabbitmq-auth-backend-ldap-15/test/src/rabbit_auth_backend_ldap_test.erl#L70).

michaelklishin commented:
@Dzol if this can be done in 2-3 days, we can do it in this issue.

Dzol commented:
@michaelklishin: I should be able to do this by the end of the day.

Commit:
DO NOT MERGE: Make tag query tests more like integration/system tests

dcorbacho commented on test/src/rabbit_auth_backend_ldap_test.erl:
```diff
@@ -89,33 +89,64 @@ internal_followed_ldap_and_internal_test_() ->
+internal_authorization_teardown() ->
+    ok = control_action(delete_user, ["Edward"]).
+
+internal_authorization_tags() ->
```
Is it really empty? Then delete it

Dzol replied:
@dcorbacho, @hairyhum, @michaelklishin: The empty list was the easiest way to highlight an issue.

If my understanding is correct, then `[{rabbit_auth_backend_ldap, rabbit_auth_backend_internal}]` and `[rabbit_auth_backend_internal, {rabbit_auth_backend_ldap, rabbit_auth_backend_internal}]` should do **authorization** only through the internal database.

The [tags](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/rabbitmq-auth-backend-ldap-15/test/src/rabbit_auth_backend_ldap_test.erl#L103) `normal` and `monitor` are [configured](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/rabbitmq-auth-backend-ldap-15/test/src/rabbit_auth_backend_ldap_test.erl#L128-L139) only through LDAP (authorization).

The tests with these two backend configurations should fail but they succeed:

https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/rabbitmq-auth-backend-ldap-15/test/src/rabbit_auth_backend_ldap_test.erl#L109-L110

They should fail because the way they're configured should only consult the internal database for tags/authorization but it seems to consult LDAP regardless.

hairyhum replied:
@Dzol  User login authentication and authorization tags are being merged from both backends in case of `{rabbit_auth_backend_ldap, rabbit_auth_backend_internal}`

dcorbacho commented on test/src/rabbit_auth_backend_ldap_test.erl:
```diff
@@ -89,33 +89,64 @@ internal_followed_ldap_and_internal_test_() ->
-         %% corresponding query should succeed.
-         "LDAP Tag attribution", tag_check(<<"Edward">>, <<"password">>, [monitor, normal])}]
-    }.
+    [{setup,
```
Use foreachx fixture, it does what you need.

```
   [{foreachx,
      fun (ldap_only) ->
              ok = application:set_env(rabbit, auth_backends, [rabbit_auth_backend_ldap]),
              set_env(tag_query_configuration());
          (ldap_and_internal) ->
              ok = application:set_env(rabbit, auth_backends, [{rabbit_auth_backend_ldap, rabbit_auth_backend_internal}]),
              internal_authorization_setup(),
      end,
      fun (ldap_only, _) ->
              set_env(base_login_env()),
              ok = application:unset_env(rabbit, auth_backends);
          (ldap_and_internal, _) ->
              internal_authorization_teardown(),
              ok = application:unset_env(rabbit, auth_backends)
      end,
      [{ldap_only, fun(_, _) ->
                   {"LDAP Tag attribution", tag_check(<<"Edward">>, <<"password">>, [monitor, normal])}
           end},
       {ldap_and_internal, fun(_, _) ->
                   {"LDAP&Internal Tag attribution", tag_check(<<"Edward">>, <<"password">>, [monitor, normal])}
           end}]}].
```

Commit:
Test tag attribution with more suitable fixtures

These make testing against different backend configurations (for
access control) easier to read.

Dzol commented:
@hairyhum, @michaelklishin, & @dcorbacho: This is ready with regard to changes suggested [here](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/pull/28#issuecomment-210549204).

Commit:
Merge branch 'stable' into rabbitmq-auth-backend-ldap-15
